cmd/fs-v1-multipart.go

```diff
@@ -18,324 +18,190 @@ package cmd
+	partNumber, err = strconv.Atoi(result[0])
 	if err != nil {
-		if err == errFileNotFound || err == errFileAccessDenied {
-			return nil, true, nil
-		}
-		return nil, false, errors.Trace(err)
-	}
-	defer fs.rwPool.Close(pathJoin(fs.fsPath, minioMetaMultipartBucket, uploadsPath))
-
-	// Read `uploads.json`.
-	uploadIDs := uploadsV1{}
-	if _, err = uploadIDs.ReadFrom(rlk.LockedFile); err != nil {
-		return nil, false, err
+		return 0, "", errUnexpected
 	}
+	return partNumber, result[1], nil
+}
 
-	index := 0
-	if uploadIDMarker != "" {
-		for ; index < len(uploadIDs.Uploads); index++ {
-			if uploadIDs.Uploads[index].UploadID == uploadIDMarker {
-				// Skip the uploadID as it would already be listed in previous listing.
-				index++
-				break
-			}
+// Appends parts to an appendFile sequentially.
+func (fs *fsObjects) backgroundAppend(bucket, object, uploadID string) {
+	fs.appendFileMapMu.Lock()
+	file := fs.appendFileMap[uploadID]
+	if file == nil {
+		file = &fsAppendFile{
+			filePath: pathJoin(fs.fsPath, minioMetaTmpBucket, fs.fsUUID, fmt.Sprintf("%s.%s", uploadID, mustGetUUID())),
 		}
+		fs.appendFileMap[uploadID] = file
 	}
+	fs.appendFileMapMu.Unlock()
 
-	for index < len(uploadIDs.Uploads) {
-		uploads = append(uploads, MultipartInfo{
-			Object:    objectName,
-			UploadID:  uploadIDs.Uploads[index].UploadID,
-			Initiated: uploadIDs.Uploads[index].Initiated,
-		})
-		count--
-		index++
-		if count == 0 {
-			break
-		}
-	}
+	file.Lock()
+	defer file.Unlock()
 
-	end := (index == len(uploadIDs.Uploads))
-	return uploads, end, nil
-}
+	// Since we append sequentially nextPartNumber will always be len(file.parts)+1
+	nextPartNumber := len(file.parts) + 1
+	uploadIDDir := fs.getUploadIDDir(bucket, object, uploadID)
 
-// listMultipartUploadsCleanup - lists all multipart uploads. Called by fs.cleanupStaleMultipartUpload()
-func (fs fsObjects) listMultipartUploadsCleanup(bucket, prefix, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (lmi ListMultipartsInfo, e error) {
-	result := ListMultipartsInfo{}
-	recursive := true
-	if delimiter == slashSeparator {
-		recursive = false
+	entries, err := readDir(uploadIDDir)
+	if err != nil {
+		errorIf(err, "error reading directory %s", uploadIDDir)
+		return
 	}
+	sort.Strings(entries)
 
-	result.IsTruncated = true
-	result.MaxUploads = maxUploads
-	result.KeyMarker = keyMarker
-	result.Prefix = prefix
-	result.Delimiter = delimiter
-
-	// Not using path.Join() as it strips off the trailing '/'.
-	multipartPrefixPath := pathJoin(bucket, prefix)
-	if prefix == "" {
-		// Should have a trailing "/" if prefix is ""
-		// For ex. multipartPrefixPath should be "multipart/bucket/" if prefix is ""
-		multipartPrefixPath += slashSeparator
-	}
-	multipartMarkerPath := ""
-	if keyMarker != "" {
-		multipartMarkerPath = pathJoin(bucket, keyMarker)
+	lenEntries := len(entries)
+	if lenEntries == 0 {
+		errorIf(errUnexpected, "%s is empty", uploadIDDir)
```

Comment:
We don't need to do this i think, since entries range would be skipped and returned quickly. Also there is no reason to print an error lo here..

Reply:
I added it to log it as we should never be in this situation. But we dont lose much by removing this error message too, because I think we will never hit this case. One possible situation where we will hit this case is when cleanup has cleaned the uploadID and not deleted the directory and just then if backgroundAppend sees 0 entries. This is a highly improbable case,so I will delete this logging.